pkg/urbit/jets/e/slaw.c

```diff
@@ -0,0 +1,508 @@
+    u3a_free(c);
+
+    if (_(u3a_is_atom(m))) {
+      return 0;
+    }
+
+    u3_atom raw = u3k(u3t(m));
+    u3z(m);
+
+    u3_noun hok = u3j_cook("u3we_slaw_fynd_p", u3k(cor), "fynd");
+    return u3nc(0, u3n_slam_on(hok, u3k(raw)));
+  }
+
+  // There must now be a - or it is invalid.
+  CONSUME('-');
+
+  // The next possible case is a "short" moon. (~ab-cd-ef)
+  TRY_GET_SYLLABLE(e);
+
+  ENSURE_NOT_END();
+
+  TRY_GET_SYLLABLE(f);
+
+  if (*cur == 0) {
+    u3_noun a_part = po_find_prefix(a_one, a_two, a_three);
+    u3_noun b_part = po_find_suffix(b_one, b_two, b_three);
+    u3_noun c_part = po_find_prefix(c_one, c_two, c_three);
+    u3_noun d_part = po_find_suffix(d_one, d_two, d_three);
+    u3_noun e_part = po_find_prefix(e_one, e_two, e_three);
+    u3_noun f_part = po_find_suffix(f_one, f_two, f_three);
+
+    u3_noun m = combine(f_part, combine(e_part, combine(d_part,
+                combine(c_part, combine(b_part, a_part)))));
+    u3a_free(c);
+
+    if (_(u3a_is_atom(m))) {
+      return 0;
+    }
+
+    u3_atom raw = u3k(u3t(m));
+    u3z(m);
+    u3_noun hok = u3j_cook("u3we_slaw_fynd_p", u3k(cor), "fynd");
+    return u3nc(0, u3n_slam_on(hok, u3k(raw)));
+  }
+
+  // There must now be a - or it is invalid.
+  CONSUME('-');
+
+  // The next possible case is a "long" moon. (~ab-cd-ef-gh)
+  TRY_GET_SYLLABLE(g);
+
+  ENSURE_NOT_END();
+
+  TRY_GET_SYLLABLE(h);
+
+  if (*cur == 0) {
+    u3_noun a_part = po_find_prefix(a_one, a_two, a_three);
+    u3_noun b_part = po_find_suffix(b_one, b_two, b_three);
+    u3_noun c_part = po_find_prefix(c_one, c_two, c_three);
+    u3_noun d_part = po_find_suffix(d_one, d_two, d_three);
+    u3_noun e_part = po_find_prefix(e_one, e_two, e_three);
+    u3_noun f_part = po_find_suffix(f_one, f_two, f_three);
+    u3_noun g_part = po_find_prefix(g_one, g_two, g_three);
+    u3_noun h_part = po_find_suffix(h_one, h_two, h_three);
+
+    u3_noun m = combine(h_part, combine(g_part, combine(f_part,
+                combine(e_part, combine(d_part, combine(c_part,
+                combine(b_part, a_part)))))));
+    u3a_free(c);
+
+    if (_(u3a_is_atom(m))) {
+      return 0;
+    }
+
+    u3_atom raw = u3k(u3t(m));
+    u3z(m);
+    u3_noun hok = u3j_cook("u3we_slaw_fynd_p", u3k(cor), "fynd");
+    return u3nc(0, u3n_slam_on(hok, u3k(raw)));
+  }
+
+  // At this point, the only thing it could be is a long comet, of the form
+  // ~ab-cd-ef-gh--ij-kl-mn-op
+
+  CONSUME('-');
+  CONSUME('-');
+
+  TRY_GET_SYLLABLE(i);
+  ENSURE_NOT_END();
+  TRY_GET_SYLLABLE(j);
+  CONSUME('-');
+  TRY_GET_SYLLABLE(k);
+  ENSURE_NOT_END();
+  TRY_GET_SYLLABLE(l);
+  CONSUME('-');
+  TRY_GET_SYLLABLE(m);
+  ENSURE_NOT_END();
+  TRY_GET_SYLLABLE(n);
+  CONSUME('-');
+  TRY_GET_SYLLABLE(o);
+  ENSURE_NOT_END();
+  TRY_GET_SYLLABLE(p);
+
+  if (*cur != 0) {
+    // We've parsed all of a comet shape, and there's still more in the
+    // string. Error.
+    u3a_free(c);
+    return u3_none;
+  }
```

Comment:
I missed this yield case earlier, this looks right and should like us parse arbitrary-length `@p`. My bad!

Reply:
I actually changed this in response to your previous request. I've changed the comment from "Error" to "Bail to the interpreter" instead.